### ISSUE TITLE
Apply u2net to cell instance segmentation

### ISSUE THREAD

**Issue by FatmaMazen2021** [Akhilesh64/Image-Segmentation-U-2-Net#8]
Dear sir
Thank you for this simple and great implementation of the u2net architecture
I work in a cell instance segmentation project . I have prepared my data in two folders, the first one is the cell images the second one is the mask PNG images created from rle annotations
I have run the training for 1000 epochs but when I want to predict on test image I got black image I can not see the mask

**Comment by Akhilesh64** [comment]
@FatmaMazen2021 Can you share a sample image and its annotation ?

**Comment by FatmaMazen2021** [comment]
Thank you for your fast response
This is the competition data
https://www.kaggle.com/c/sartorius-cell-instance-segmentation/data
And here I have converted rle annotations into PNG images 
https://www.kaggle.com/fatmamazen/sartorius-png-masks
I don't know why they look black but when I use pyplot.imshow, mask is correctly shown
I have a question about dimensionality of train and mask images. Should both be grayscale or the train image is grayscale and mask is grayscale?

**Comment by Akhilesh64** [comment]
The input training images are assumed to be RGB [see here](https://github.com/Akhilesh64/Image-Segmentation-U-2-Net/blob/eef9854d505abb3a79aa2dca7b5aae359141a5f4/U2-net.py#L321). Using the predict.py file you should get grayscale frames of all the 6 outputs of the model which you can combine to get a final grayscale mask.

> I don't know why they look black but when I use pyplot.imshow, mask is correctly shown

About this, the masks are normalized between 0-1, that's why you get a blank mask, multiply the image by 255  and  you should be able to see the results.

**Comment by FatmaMazen2021** [comment]
Thank you for your reply
Could you kindly tell me which function I should use to combine the output and where should I include it?
Will it be something thing like
Overall=preds[0]+preds[1]+preds[2]+......+preds[6]?
Or overall=np.array(preds)
Overall_img=Image.fromarray(overall)*255?
Another question is about thresholding,is it an option here?
Could you kindly tell me how to show the overall output grayscale mask? 
I am required to further convert it to rle coding not to save it as PNG image. Should I multiply the output mask by 255 before converting it to rle?
I will be grateful if you help me with this part 
Thanks in advance.

**Comment by Akhilesh64** [comment]
I forgot the model outputs 7 masks. preds[0] is actually the combination of all the masks, so preds[0]*255 should work as the final mask. Thresholding is done to improve the resultant mask obtained. All this can be found in the predict.py file. I dont know about rle encoding so I wont be able to help you in that matter.

**Comment by FatmaMazen2021** [comment]
I have figured out the reason for getting black mask. The initial masks are normalized between 0 and 1 and the function getData divide the pixels by 255 which means no pixel of the input image will be 1 all of them will be 0 due to very small value. I have commented this line, trained model for small number of epochs just to see the effect of removing this line. I have seen masks for the first time😀😀
But I don't know why I can see them in blue and green colors
Not as grayscale images.should I save them first to PNG image and then read them and show them?
Thank you very much for your code and your support

**Comment by FatmaMazen2021** [comment]
Could you kindly share the predict file of this example with me?

**Comment by Akhilesh64** [comment]
You can find the prediction file in the repo itself.

**Comment by FatmaMazen2021** [comment]
Could you kindly tell me why you replaced load_img by Image.open in predict.py?
If I want to use the combined mask as my final mask 
img = load_img(fn)
    img = img.resize((256,256))
    img = img_to_array(img)
    img = np.expand_dims(img,axis=0)
    model = load_model('../input/u2net-model-27-11/model_27_11.h5', compile=False)
    preds = model.predict(img)
    preds = np.squeeze(preds)
    shape = preds[0,:,:].shape
    frame = binarize(preds[1,:,:], threshold = 0.5)
    #print(frame)
    frame1 = np.reshape(frame,(shape[0], shape[1]))
    frame2 = cv2.resize(frame1, (704,520), interpolation = cv2.INTER_AREA)
    frame2 = np.expand_dims(np.array(frame2),axis=-1)
Is this the right sequence of extracting the combined mask?
Thanks in advance

**Comment by FatmaMazen2021** [comment]
Is there a way to get the mask of each segmented object in a separate image?

**Comment by Akhilesh64** [comment]
@FatmaMazen2021 Sorry for the late reply. I was stuck up in some other work. 

> Could you kindly tell me why you replaced load_img by Image.open in predict.py? 

Both are same Tensorflow uses PIL to load Images so there's no difference.

> If I want to use the combined mask as my final mask

You dont have to combine masks yourself. preds[0] is the combined mask output by the model.

> Is there a way to get the mask of each segmented object in a separate image?

U2-Net is a binary segmentation network. So I dont think this is possible with the current architecture as the problem you state looks like instance or semantic segmentation.

**Comment by FatmaMazen2021** [comment]
> combined

Is this the right code ?
Another question is that if I want to use preds[0], should I use this line before or after np.squeeze?
I want to understand the idea behind np.squeeze
Yes my problem is instance segmentation problem
I have used opencv to find a list of instances 
But I still have problem in submission

**Comment by FatmaMazen2021** [comment]
> 

Thank you very much for your help and support

**Comment by FatmaMazen2021** [comment]
def post_process(probability, threshold=0.5, min_size=300):
    mask = cv2.threshold(probability, threshold, 1, cv2.THRESH_BINARY)[1]
    num_component, component = cv2.connectedComponents(mask.astype(np.uint8))
    predictions = []
    for c in range(1, num_component):
        p = (component == c)
        if p.sum() > min_size:
            a_prediction = np.zeros((520, 704), np.float32)
            a_prediction[p] = 1
            predictions.append(a_prediction)
    return predictions

I have used this function to get list of instances

**Comment by FatmaMazen2021** [comment]
Could you modify the architecture to support instance segmentation?

**Comment by FatmaMazen2021** [comment]
Do you think that
 replacing sigmoid function with softmax and
 binary crossentropy by categorical crossentropy
And loading data as multiclass using image data generator will solve the problem?

**Comment by Akhilesh64** [comment]
@FatmaMazen2021 Yes this could solve this but I am not sure. Give it a try

**Comment by FatmaMazen2021** [comment]
> @FatmaMazen2021 Yes this could solve this but I am not sure. Give it a try

I have actually tried it ,but I do not know where to add number of classes in the output layer?
And where to get instance masks?
I will be grateful if you could help me with this part 
If I manage to get good results,may be will publish a great paper.It will be a novel approach to use u2net in instance segmentation. All of your work will be cited in my next paper

**Comment by FatmaMazen2021** [comment]
If you don't mind, I may share the modifications with you and discuss this issue to get a novel architecture for instance segmentation

**Comment by Akhilesh64** [comment]
Could you share a colab demo of your work with modifications over my mail. I'll be able to get a better understanding and we could continue this over mail without spamming this section.

**Comment by FatmaMazen2021** [comment]
Could you kindly share your mail with me?

**Comment by Akhilesh64** [comment]
My email and linkedin profile are listed on my main github profile page.

**Comment by FatmaMazen2021** [comment]
> My email and linkedin profile are listed on my main github profile page.

Done. kindly check your email
